File: frontend/src/services/budgets-service.ts
import { API_CONFIG } from '@/config/constants';
import type { Budget, BudgetFormData, BudgetHistory, BudgetStatus } from '@/types';

import { apiClient } from './api-client';
import { BaseService } from './base-service';

class BudgetsService extends BaseService<Budget, BudgetFormData, BudgetFormData> {
  constructor() {
    super(API_CONFIG.ENDPOINTS.BUDGETS);
  }

  async getStatus(params?: { month?: number; year?: number }): Promise<BudgetStatus[]> {
    return apiClient.get<BudgetStatus[]>(API_CONFIG.ENDPOINTS.BUDGETS_STATUS, params);
  }

  async getHistory(params: {
    category: string;
    months?: number;
  }): Promise<BudgetHistory[]> {
    return apiClient.get<BudgetHistory[]>(API_CONFIG.ENDPOINTS.BUDGET_HISTORY, params);
  }
}

export const budgetsService = new BudgetsService();

File: apps/mobile/lib/services/budgets_service.dart
import '../models/budget.dart';
import 'base_service.dart';

class BudgetsService extends BaseService<Budget> {
  BudgetsService(super.client)
      : super(
          resourcePath: '/api/v1/budgets/',
          fromJson: Budget.fromJson,
          toJson: (b) =>
              {'category': b.category, 'limit_amount': b.limitAmount},
        );

  Future<List<BudgetStatus>> status(int month, int year) async {
    final r = await client.dio.get<dynamic>(
      '${resourcePath}status/',
      queryParameters: {'month': month, 'year': year},
    );
    if ((r.statusCode ?? 0) >= 400) throw ApiException(r.statusCode, r.data);
    final data = r.data;
    final list = data is Map ? (data['results'] ?? const []) : data;
    return (list as List)
        .map((e) => BudgetStatus.fromJson(e as Map<String, dynamic>))
        .toList();
  }
}
